File: 프로그래머스/1/118666. 성격 유형 검사하기/성격 유형 검사하기.java
import java.util.*;

class Solution {
    public String solution(String[] survey, int[] choices) {
        StringBuilder answer = new StringBuilder();
        
        Character[] types = {'R','T','C','F','J','M','A','N'};
        Map<Character, Integer> map = new HashMap<>();
        
        for (char i : types) {
            map.put(i, 0);
        }
        
        for (int i = 0; i < survey.length; i++) {
            char left = survey[i].charAt(0);
            char right = survey[i].charAt(1);
            int choice = choices[i];
            
            if (choice < 4) map.put(left, map.get(left) + (4-choice));
            else map.put(right, map.get(right) + (choice-4));
        }
        
        for (int i = 0; i < types.length; i+=2) {
            if (map.get(types[i]) >= map.get(types[i+1])) answer.append(types[i]);
            else answer.append(types[i+1]);
        }
        
        
        return answer.toString();
    }
}

File: 프로그래머스/1/118666. 성격 유형 검사하기/성격 유형 검사하기.py
def solution(survey, choices):
    answer = ''
    
    types = 'RTCFJMAN'
    index = {}
    
    for t in types:
        index[t] = 0
        
    for i in range(len(survey)):
        left = right = 0
        
        if choices[i] <= 3:
            left += (4-choices[i])
        elif choices[i] >= 5:
            right += (choices[i]-4)
            
        if left > right:
            index[survey[i][0]] += left
        elif left < right:
            index[survey[i][1]] += right
            
    for i in range(0, len(types), 2):
        if index[types[i]] > index[types[i+1]]:
            answer += types[i]
        elif index[types[i]] == index[types[i+1]]:
            answer += min(types[i], types[i+1])
        else:
            answer += types[i+1]
            
        
    return answer